Notify user of connection issue.

godofgrunts opened the pull request:
PR for #14 

You shouldn't accept this PR as is, I wanted to show two different options and discuss about which version would be better.

In `obs_websocket.gd` I added a new if conditional `if json_response.has("error")` in the `_on_data_received` function.

You could just send a generic signal with the error and let the client handle it, or you could make a huge match statement with common errors if you want to handle it yourself.

As it is now, the `obs_websocket.gd` is sending a generic error and the "client" `obs_ui.gd` is handling what to do with that information.

you-win commented:
I am in favor of letting the client handle errors and keeping `obs_websocket.gd` fairly lean, as I think it would be a big undertaking to handle every single error.

Disclosure: I mainly use `obs_websocket.gd` as a library and created the editor plugin as a demo, which is why the plugin is a bit buggy 🙃

godofgrunts commented:
Yeah I think I agree with you. I prefer the idea of letting the client handle it as well so I'll stick with just emitting an error signal.

godofgrunts commented:
Okay so I removed `obs_websocket.gd` from trying to match stuff itself. I also added the `emit_signal` to `_on_connection_error` because I kept running into that problem while opening and closing OBS during testing.

Moved the error handling to the bottom of `obs_ui.gd` so it didn't get lost in the other parts of the code. I also added more comments explaining what is going on.

I also am creating the popup in code now. Previously it wasn't working in the tool portion of the UI, but it is now.

This commit is ready for review.

godofgrunts commented on addons/obs_websocket_gd/obs_websocket.gd:
```diff
@@ -163,6 +165,10 @@ func _on_data_received() -> void:
+		
+	if json_response.has("error"):
+		print("Error: %s" % json_response["error"])
+		emit_signal("obs_error", json_response["error"])
```
Hmm. I'm not sure. If we do that are we basically treating every error as fatal?

you-win replied:
Yes. I guess it wouldn't hurt to also broadcast the entire error json in the `obs_updated`. At this point, I'm good with either approach.

godofgrunts replied:
Since we already decided to let the client handle the errors I think it's probably safest to allow the obs_updated signal to occur, but let me go over the code a bit more and see if I can see a possible solution that I didn't see before.

godofgrunts replied:
Okay so we may not need to create a new signal like this. The client already is receiving the `obs_update` signal which would include the error the client really could just handle it in it's `_on_obs_updated` right?

I may have just created extra work haha.

Let me try this out and I'll  get back with you.

godofgrunts replied:
Okay I figured out a better way to do it using the signal we already had. That way the user doesn't have to worry about different types of signals.

godofgrunts commented:
And of course I forgot to fix the comment section.

you-win commented:
Everything looks good to me. Is this ready to merge?

godofgrunts commented:
> Everything looks good to me. Is this ready to merge?

Yeah I think so.

you-win commented:
Thanks for the PR and thanks for being open to discussion!

godofgrunts commented:
Thanks for putting up with my multiple PRs! :D
